src/main/java/org/broadinstitute/hellbender/engine/GATKPathSpecifier.java

```diff
@@ -0,0 +1,162 @@
+package org.broadinstitute.hellbender.engine;
+
+import com.google.cloud.storage.contrib.nio.CloudStorageFileSystem;
+import org.broadinstitute.barclay.argparser.TaggedArgument;
+import org.broadinstitute.barclay.argparser.TaggedArgumentParser;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.gcs.BucketUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * GATK tool command line arguments that are input or output resources. These can
+ * have an optional name supplied on the command line, as well as one or more optional tag/value pairs.
+ */
+public class GATKPathSpecifier extends PathSpecifier implements TaggedArgument, Serializable {
+    private static final long serialVersionUID = 1L;
+
+    public static final String GCS_SCHEME = "gs";
+    public static final String HDFS_SCHEME = "hdfs";
+
+    private String tagName;
+    private Map<String, String> tagAttributes;
+
+    /**
+     * The raw value for this specifier as provided by the user. Can ba a local file or valid URI.
+     * @param uriString
+     */
+    public GATKPathSpecifier(final String uriString) {
+        super(uriString);
+    }
+
+    @Override
+    public Path toPath() {
+        // special case GCS, in case the filesystem provider wasn't installed properly but is available.
+        if (CloudStorageFileSystem.URI_SCHEME.equals(getURI().getScheme())) {
+            final Path tempPath = BucketUtils.getPathOnGcs(getURIString());
+            setCachedPath(tempPath);
+            return tempPath;
+        } else {
+            return super.toPath();
+        }
+    }
+
+    //TODO: should this wrap the stream in a .gz in a BlockCompressedInputStream like IOUtils does?
+    @Override
+    public InputStream getInputStream() {
+        if (!isPath()) {
+            throw new UserException(getToPathFailureReason());
+        }
+
+        try {
+            InputStream inputStream;
+            if (getURI().getScheme().equals(GCS_SCHEME)) {
+                Path p = BucketUtils.getPathOnGcs(getURIString());
+                inputStream = Files.newInputStream(p);
+            } else if (getURI().getScheme().equals(HDFS_SCHEME)) {
+                org.apache.hadoop.fs.Path file = new org.apache.hadoop.fs.Path(getURIString());
```

Comment:
Good question. I included this because its what `BucketUtils.openFile` does (and thus what I needed for my bigger version of this branch where I used the URI class for all reads and reference inputs). But maybe it isn't necessary ? Somehow I thought the jsr203hadoop provider was flake, but I'm not sure where I got that from.

Reply:
Good question. I included this because its what `BucketUtils.openFile` does (and thus what I needed for my bigger version of this branch where I used the URI class for all reads and reference inputs). But maybe it isn't necessary ? Somehow I thought the jsr203hadoop provider was flaky, but I'm not sure where I got that from.

Reply:
It might be flakey, we haven't done any sort of stress testing on it and it's not really maintained as far as I know.  My guess is that's it's fine for getting a stream though.